jedi/inference/gradual/annotation.py

```diff
@@ -362,45 +442,73 @@ def iter(iterable: Iterable[_T]) -> Iterator[_T]: ...
+            annotation_generics = annotation_value.get_generics()
+            tuple_annotation, = annotation_value.execute_annotation()
+            # TODO: is can we avoid using this private method?
+            if tuple_annotation._is_homogenous():
+                # The parameter annotation is of the form `Tuple[T, ...]`,
+                # so we treat the incoming tuple like a iterable sequence
+                # rather than a positional container of elements.
+                for nested_annotation_value in annotation_generics[0]:
+                    _merge_type_var_dicts(
+                        type_var_dict,
+                        _infer_type_vars(
+                            nested_annotation_value,
+                            value_set.merge_types_of_iterate(),
+                        ),
+                    )
+            else:
+                # The parameter annotation has only explicit type parameters
+                # (e.g: `Tuple[T]`, `Tuple[T, U]`, `Tuple[T, U, V]`, etc.) so we
+                # treat the incoming values as needing to match the annotation
+                # exactly, just as we would for non-tuple annotations.
+
+                for element in value_set:
+                    py_class = element.get_annotated_class_object()
+                    if not isinstance(py_class, GenericClass):
+                        py_class = element
+
+                    if not isinstance(py_class, DefineGenericBase):
+                        continue
+
+                    merge_pairwise_generics(annotation_value, py_class)
+
     elif isinstance(annotation_value, GenericClass):
-        name = annotation_value.py__name__()
-        if name == 'Iterable':
+        if annotation_name == 'Iterable' and not is_class_value:
             given = annotation_value.get_generics()
             if given:
                 for nested_annotation_value in given[0]:
                     _merge_type_var_dicts(
                         type_var_dict,
                         _infer_type_vars(
                             nested_annotation_value,
-                            value_set.merge_types_of_iterate()
+                            value_set.merge_types_of_iterate(),
                         )
                     )
-        elif name == 'Mapping':
-            given = annotation_value.get_generics()
-            if len(given) == 2:
-                for value in value_set:
-                    try:
-                        method = value.get_mapping_item_values
-                    except AttributeError:
-                        continue
-                    key_values, value_values = method()
+        else:
+            # Note: we need to handle the MRO _in order_, so we need to extract
+            # the elements from the set first, then handle them, even if we put
+            # them back in a set afterwards.
+            for element in value_set:
+                if not hasattr(element, 'is_instance'):
+                    continue
```

Comment:
Should not be necessary. If it's not available that's a bug.

Reply:
da9d3121 :thumbsup: